src/components/cube.jsx

```diff
@@ -1,20 +1,58 @@
-import React, { useRef, useState } from 'react';
+import React,
+{
+  forwardRef,
+  useImperativeHandle,
+  useRef,
+  useState,
+} from 'react';
 import PropTypes from 'prop-types';
+import copy from 'copy-to-clipboard';
 
 import useColumnOptions from './use/column-options';
 import CubeColumnChooser from './cube-column-chooser';
 import HypercubeTable from './hypercube-table';
 import useForce from './use/force';
+import useModel from './use/model';
 
 import './cube.pcss';
 
-export default function Cube({ app, tableData: { initialColumns }, closeOnClickOutside }) {
+// The component needs to be wrapped in `forwardRef` to give access to the
+// ref object assigned using the `ref` prop.
+const Cube = forwardRef(({ app, tableData: { initialColumns }, closeOnClickOutside }, ref) => {
   const selectableColumns = useColumnOptions(app);
   const [columns, setColumns] = useState(initialColumns);
   const currentHeader = useRef(null);
   const columnToReplace = useRef(null);
   const addOpen = useRef(false);
   const forceUpdate = useForce();
+  let model = null;
+  let hypercubeProps = null;
+
+  // Any instance of the component is extended with what is returned from the
+  // callback passed as the second argument.
+  useImperativeHandle(ref, () => ({
+    copyToClipboard() {
+      if (hypercubeProps) {
+        copy(JSON.stringify(hypercubeProps));
+      }
+    },
+    async exportHypercube() {
+      if (model) {
+        const result = await model.exportData('OOXML', '/qHyperCubeDef');
+        const engineUrl = new URLSearchParams(document.location.search).get('engine_url');
+        if (engineUrl) {
+          const wsUrl = new URL(engineUrl);
+          const protocol = wsUrl.protocol === 'wss' ? 'https' : 'http';
+          const elem = document.createElement('a');
+          elem.href = `${protocol}://${wsUrl.host}${result.qUrl}`;
```

Comment:
You could do `wsUrl.origin.replace('http', 'ws')` instead, e.g.:

```
new URL("wss://foo.bar/baz?que=qux").origin.replace('http', 'ws')
"wss://foo.bar"
```

Reply:
smooth!